[FEATURE] Ne donner accès à l'espace surveillant qu'à certains centres de certification (PIX-3770).

aceol opened the pull request:
## :christmas_tree: Problème
Le toggle de fin de test se fait pour tout PIX, hors on souhaiterais pouvoir le faire uniquement pour certains centre de certification

## :gift: Solution
Ajouter une variable d'environnement FT_ALLOWED_CERTIFICATION_CENTER_IDS_FOR_END_TEST_SCREEN_REMOVAL qui renseigne une liste d'id de centre de certification.
Mettre le toggle en fonction d'un centre certification donné (a partir de session.certificationCenterId)

TODO voir ce qui doit être géré coté front:
- affichage de la page de fin de test: on peut utiliser les [meta de json-api](https://jsonapi.org/format/#document-meta)
- Affichage du mot de passe surveillant coté certif supervisorPasswordShouldBeDisplayed

## :star2: Remarques
Il y a 122 centres de certification sans external id en prod.

Dans une autre PR:
- gestion du lien en menu certif en fonction du centre de l'utilisateur
- ecran de fin de test (fin de certification)

## :santa: Pour tester
Tester pour toutes ces routes que la feature n'est pas activée si END_TEST_SCREEN_REMOVAL_WHITELIST ne contient pas l'identifiant du certification center
- POST /api/certification-candidates/:id/authorize-to-start
- POST /api/certification-candidates/:id/authorize-to-resume
- PATCH /api/certification-candidates/{id}/end-assessment-by-supervisor
- GET /api/sessions/121/supervising
- POST /api/sessions/supervise
- POST /api/certification-courses (save)
- /api/token (login in certif)

aceol committed:
tech(api) Add FT_ALLOWED_CERTIFICATION_CENTER_IDS_FOR_END_TEST_SCREEN_REMOVAL

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr3875.review.pix.fr
- App (.org): https://app-pr3875.review.pix.org
- Orga: https://orga-pr3875.review.pix.fr
- Certif: https://certif-pr3875.review.pix.fr
- Admin: https://admin-pr3875.review.pix.fr
- API: https://api-pr3875.review.pix.fr/api/

 Please check it out!

aceol committed:
refacto(api) Fix getArrayOfStrings when varenv is not set

aceol committed:
feat(api) Add service for isEndTestScreenRemovalEnabledBySessionId

aceol committed:
feat(api) Add service and repo for isEndTestScreenRemovalEnabledByCandidateId

aceol committed:
:sparkles: feat(api) Feature toggle remove end test screen depends on certification center for certification-candidate-controller

aceol committed:
:sparkles: feat(api) Feature toggle remove end test screen depends on certification center for session-for-supervising-controller

aceol committed:
:recycle: refacto(api): rename controller

aceol committed:
:sparkles: feat(api) Feature toggle remove end test screen depends on certification center for certification-course #POST

aceol committed:
:sparkles: feat(api) Feature toggle remove end test screen depends on certification center to authenticate user

aceol committed:
:recycle: refacto(api) Always serialize session with supervisor-password

Even if the feature flag endTestScreen is not enabled, we can provide the supervisor password, as it's either empty or useless

aceol commented on api/lib/config.js:
```diff
@@ -162,6 +162,9 @@ module.exports = (function () {
       isEmailValidationEnabled: isFeatureEnabled(process.env.FT_VALIDATE_EMAIL),
       isManageUncompletedCertifEnabled: isFeatureEnabled(process.env.FT_MANAGE_UNCOMPLETED_CERTIF_ENABLED),
       isEndTestScreenRemovalEnabled: isFeatureEnabled(process.env.FT_END_TEST_SCREEN_REMOVAL_ENABLED),
+      isEndTestScreenRemovalEnabledCertificationCenterIds: _getArrayOfStrings(
```
```suggestion
      allowedCertificationCenterIdsForEndTestScreenRemoval: _getArrayOfStrings(
```

aceol commented on api/tests/integration/infrastructure/repositories/end-test-screen-removal-repository_test.js:
```diff
@@ -0,0 +1,84 @@
+            await databaseBuilder.commit();
+            sinon
+              .stub(featureToggles, 'isEndTestScreenRemovalEnabledCertificationCenterIds')
+              .value([certificationCenterId]);
```
```suggestion
              .value([0, certificationCenterId, 1]);
```

aceol commented on api/lib/application/certification-candidates/certification-candidates-controller.js:
```diff
@@ -1,16 +1,20 @@
 
 module.exports = {
   async authorizeToStart(request, h) {
-    if (!featureToggles.isEndTestScreenRemovalEnabled) {
```
suggestion: deplacer dans le pre-handler?

aceol commented on api/lib/application/certification-candidates/certification-candidates-controller.js:
```diff
@@ -20,11 +24,15 @@ module.exports = {
 
   async authorizeToResume(request, h) {
-    if (!featureToggles.isEndTestScreenRemovalEnabled) {
+    const certificationCandidateId = request.params.id;
```
suggestion: deplacer dans le pre-handler?

aceol replied:
rajouter aussi pour endAssessment

aceol commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -39,7 +40,10 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
-        sinon.stub(featureToggles, 'isEndTestScreenRemovalEnabled').value(false);
+        sinon.stub(endTestScreenRemovalService, 'isEndTestScreenRemovalEnabledByCandidateId');
+        endTestScreenRemovalService.isEndTestScreenRemovalEnabledByCandidateId
+          //   .withArgs(99)
```
to delete

aceol replied:
```suggestion
```

aceol commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -52,10 +56,11 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
 
         // when
-        const error = await catchErr(certificationCandidateController.authorizeToStart)(request, hFake);
+        //  const error = await catchErr(certificationCandidateController.authorizeToStart)(request, hFake);
```
```suggestion
```

aceol commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -52,10 +56,11 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
 
         // then
-        expect(error).to.be.an.instanceOf(NotFoundError);
+        //  expect(error).to.be.an.instanceOf(NotFoundError);
```
```suggestion
```

aceol commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -64,7 +69,8 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
     describe('when FT_END_TEST_SCREEN_REMOVAL_ENABLED is enabled', function () {
```
rename VAR_ENV

aceol commented on api/tests/acceptance/application/session/session-for-supervising-controller-get_test.js:
```diff
@@ -12,7 +12,7 @@ describe('Acceptance | Controller | session-for-supervising-controller-get', fun
   context('when FT_IS_END_TEST_SCREEN_REMOVAL_ENABLED is enabled', function () {
```
Rename FT

aceol committed:
:recycle: refacto(api) Do not need to check order (flaky test)

P-Jeremy commented on api/lib/config.js:
```diff
@@ -162,6 +162,9 @@ module.exports = (function () {
       isManageUncompletedCertifEnabled: isFeatureEnabled(process.env.FT_MANAGE_UNCOMPLETED_CERTIF_ENABLED),
       isEndTestScreenRemovalEnabled: isFeatureEnabled(process.env.FT_END_TEST_SCREEN_REMOVAL_ENABLED),
+      allowedCertificationCenterIdsForEndTestScreenRemoval: _getArrayOfStrings(
+        process.env.FT_ALLOWED_CERTIFICATION_CENTER_IDS_FOR_END_TEST_SCREEN_REMOVAL
```
Du coup si je comprends bien c'est une whitelist ? Là vu le nommage on dirait un feature toggle de plus non ?

aceol replied:
Deplacer 
 dans features en tant que whitelist
API renvoie featureToggles.isEndTestScreenRemovalEnabled > true si la whitelist contient plus d'1 element

aceol committed:
:sparkles: feat(certif) Feature toggle remove end test screen depends on certification center to authenticate user

The link in authenticated is not linked to a sesission/certification center, we just check that there is at least an allowed certification center

aceol commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -36,10 +37,13 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
-        sinon.stub(featureToggles, 'isEndTestScreenRemovalEnabled').value(false);
+        sinon.stub(endTestScreenRemovalService, 'isEndTestScreenRemovalEnabledByCandidateId');
+        endTestScreenRemovalService.isEndTestScreenRemovalEnabledByCandidateId
+          //   .withArgs(99)
```
A supprimer!

aceol commented on api/tests/acceptance/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -67,15 +79,16 @@ describe('Acceptance | API | Certifications candidates', function () {
   describe('PATCH /api/certification-candidates/{id}/end-assessment-by-supervisor', function () {
     context('when user is authenticated', function () {
-      describe('when FT_END_TEST_SCREEN_REMOVAL_ENABLED is enabled', function () {
+      describe('when FT_ALLOWED_CERTIFICATION_CENTER_IDS_FOR_END_TEST_SCREEN_REMOVAL is enabled for the certification center', function () {
```
Le FT n'a pas d'impact ici (pas implementé sur cette route)

aceol commented on api/lib/infrastructure/serializers/jsonapi/session-serializer.js:
```diff
@@ -5,7 +5,6 @@ const { WrongDateFormatError } = require('../../../domain/errors');
 const { isValidDate } = require('../../utils/date-utils');
```
changer le feat en refacto dans le nom du commit

P-Jeremy committed:
:recycle:refactor(api): add white list end test screen removal prehandler

aceol committed:
:recycle:refactor(api): move controller's end test screen removal check to prehandler

P-Jeremy committed:
:recycle:refactor(api): feature is enabled only if whitelist is not empty

P-Jeremy commented on api/lib/application/preHandlers/end-test-screen-removal-enabled.js:
```diff
@@ -0,0 +1,33 @@
+      sessionId
+    );
+    if (!isEndTestScreenRemovalEnabled) {
+      return h.response().code(404).takeover();
```
Renvoyer une 403 plutôt ? :thinking: 

Avec un message d'erreur explicite pour éviter ceci 

![image](https://user-images.githubusercontent.com/37305474/149121176-99ebe2d0-f5b3-4433-a635-3adf884093b7.png)

P-Jeremy committed:
:recycle:refactor(api): throw an error when certification center is not in the white list

P-Jeremy commented:
Testé fonctionnellement :+1:

clemlatz commented on api/lib/config.js:
```diff
@@ -323,5 +324,8 @@ module.exports = (function () {
-  return config;
+  return {
+    ...config,
+    getArrayOfStrings,
```
**suggestion** 
Mettre la méthode dans un fichier à part (string-utils) ? Préciser dans le nom de la fonction qu'elle change la casse ?

clemlatz commented on api/lib/infrastructure/repositories/end-test-screen-removal-repository.js:
```diff
@@ -17,6 +17,24 @@ async function isEndTestScreenRemovalEnabledBySessionId(sessionId) {
+    .select(1)
+    .from('sessions')
+    .innerJoin('certification-candidates', 'certification-candidates.sessionId', 'sessions.id')
+    .innerJoin('certification-centers', 'certification-centers.id', 'sessions.certificationCenterId')
```
**suggestion**
On peut se passer de cette jointure sur la table "certification-centers" puisqu'on a besoin juste de l'id du centre et d'aucun autre champ de cette table.

clemlatz commented on api/tests/integration/infrastructure/repositories/end-test-screen-removal-repository_test.js:
```diff
@@ -76,6 +77,86 @@ describe('Integration | Repository | EndTestScreenRemovalRepository', function (
+    });
+
+    context('the feature is enabled for the certification center associated with the given session id', function () {
+      it('returns true if isEndTestScreenRemovalEnabledCertificationCenterIds contains the id associated with given session id', async function () {
```
**suggestion**
```suggestion
      it('returns true if the session\'s certification center is whitelisted', async function () {
```

clemlatz commented on api/tests/unit/domain/services/end-test-screen-removal-service_test.js:
```diff
@@ -16,4 +16,18 @@ describe('Unit | Domain | Service | EndTestScreenRemoval', function () {
+
+  describe('#isEndTestScreenRemovalEnabledByCandidateId', function () {
+    it('should call repository with candidate id', async function () {
+      // given
+      const candidateId = Symbol('candidateId');
+      sinon.spy(endTestScreenRemovalRepository, 'isEndTestScreenRemovalEnabledByCandidateId');
+
+      // when
+      await endTestScreenRemovalService.isEndTestScreenRemovalEnabledByCandidateId(candidateId);
+
+      // then
+      expect(endTestScreenRemovalRepository.isEndTestScreenRemovalEnabledByCandidateId).to.be.calledWith(candidateId);
```
**issue**
On a remarqué deux choses : 
- si on change le comportement du composant (eg. si on commente la ligne return de la méthode du service), le test passe quand même (il faudrait plutôt tester ce que retourne la méthode, cf. https://1024pix.atlassian.net/wiki/spaces/DEV/pages/600178695/Tester)
- le repository est vraiment appelé alors qu'on est dans un test unitaire, le test passe uniquement parce que la var d'env est vide

clemlatz commented on api/lib/application/certification-candidates/certification-candidates-controller.js:
```diff
@@ -1,16 +1,20 @@
+    const isEndTestScreenRemovalEnabled = await endTestScreenRemovalService.isEndTestScreenRemovalEnabledByCandidateId(
+      certificationCandidateForSupervisingId
+    );
+    if (!isEndTestScreenRemovalEnabled) {
```
**question** 
Pourquoi ne pas faire ça dans le use-case ce qui permettrait de se passer d'un service ?

aceol replied:
C'est maintenant dans le pre-handler

octo-topi commented on api/tests/acceptance/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -37,18 +43,24 @@ describe('Acceptance | API | Certifications candidates', function () {
     context('when user is authenticated', function () {
-      describe('when FT_END_TEST_SCREEN_REMOVAL_ENABLED is enabled', function () {
+      describe('when FT_ALLOWED_CERTIFICATION_CENTER_IDS_FOR_END_TEST_SCREEN_REMOVAL is enabled for the certification center', function () {
         it('should return a 204 status code', async function () {
```
Est-il possible de tester unitairement cette erreur 204 ?

aceol replied:
Deja testé (test en "doublon" unit/acceptance)

octo-topi commented on api/tests/acceptance/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -165,8 +187,10 @@ describe('Acceptance | API | Certifications candidates', function () {
       const userId = databaseBuilder.factory.buildUser().id;
-
-      const session = databaseBuilder.factory.buildSession();
+      const certificationCenter = databaseBuilder.factory.buildCertificationCenter();
```
Quel est la RG fonctionnelle qui restreint l'accès aux certifications complémentaires sur base du FT espace surveillant ?

octo-topi commented on api/tests/acceptance/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -187,6 +211,10 @@ describe('Acceptance | API | Certifications candidates', function () {
+      sinon
+        .stub(featureToggles, 'allowedCertificationCenterIdsForEndTestScreenRemoval')
+        .value([certificationCenter.id]);
+
```
Quelle est la règle de gestion fonctionnelle qui restreint l'accès aux certifications complémentaires sur base du FT espace surveillant ? Pas connue de @asrodride

aceol replied:
Inutile ici

octo-topi commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -1,15 +1,16 @@
 describe('Unit | Controller | certifications-candidate-controller', function () {
   describe('#authorizeToStart', function () {
-    describe('when FT_END_TEST_SCREEN_REMOVAL_ENABLED is enabled', function () {
+    describe('when end screen removal is enabled', function () {
```
**suggestion**
`when associated certification center is whitelisted`

octo-topi commented on api/tests/unit/application/certification-candidates/certification-candidates-controller_test.js:
```diff
@@ -36,10 +37,13 @@ describe('Unit | Controller | certifications-candidate-controller', function ()
     });
 
-    describe('when FT_END_TEST_SCREEN_REMOVAL_ENABLED is not enabled', function () {
+    describe('when end test screen removal is not enabled', function () {
```
**suggestion**
`when associated certification center is not whitelisted`

aceol replied:
Code deplacé dans le prehandler

clemlatz commented on api/lib/domain/usecases/authenticate-user.js:
```diff
@@ -20,7 +20,9 @@ function _checkUserAccessScope(scope, user) {
 
   if (scope === apps.PIX_CERTIF.SCOPE && !user.isLinkedToCertificationCenters()) {
-    if (!featureToggles.isEndTestScreenRemovalEnabled) {
+    const isEndTestScreenEnabled =
+      endTestScreenRemovalService.isEndTestScreenRemovalEnabledForSomeCertificationCenter();
+    if (!isEndTestScreenEnabled) {
```
**suggestion**
```suggestion
    const isEndTestScreenRemovalEnabled =
      endTestScreenRemovalService.isEndTestScreenRemovalEnabledForSomeCertificationCenter();
    if (!isEndTestScreenRemovalEnabled) {
```
C'est le contraire 🙂

aceol replied:
OK on renomme

clemlatz commented on certif/app/controllers/authenticated/sessions/details/parameters.js:
```diff
@@ -22,7 +22,7 @@ export default class SessionParametersController extends Controller {
 
   get supervisorPasswordShouldBeDisplayed() {
-    return this.featureToggles.featureToggles.isEndTestScreenRemovalEnabled;
+    return Boolean(this.session.supervisorPassword);
```
**issue**
Sera toujours `true` car toutes les sessions en prod ont déjà un mot de passe surveillant. Il me semble qu'on avait fait comme ça pour ne pas avoir, le jour où on activerait le toggle, à régénérer manuellement les mots de passes des sessions crée avant l'activation.
À la place on peut utiliser la propriété de `currentUser.currentAllowedCertificationCenterAccess` comme dans #3931 (dans une autre PR peut-être ?)

aceol replied:
Le mot de passe de session n'est pas renvoyé coté back si la session n'est pas lié à un centre en whitelist [cf](https://github.com/1024pix/pix/pull/3875/commits/6baedfbe7177561524a393d63ecf30c0e920244b), ca aurait pu avoir du sens avec `currentUser.currentAllowedCertificationCenterAccess`, on pourra voir dans le ticket qui passera apres #3931 pour la correction de la gestion du FT coté front

clemlatz commented on api/lib/config.js:
```diff
@@ -163,7 +163,7 @@ module.exports = (function () {
       isEmailValidationEnabled: isFeatureEnabled(process.env.FT_VALIDATE_EMAIL),
       isManageUncompletedCertifEnabled: isFeatureEnabled(process.env.FT_MANAGE_UNCOMPLETED_CERTIF_ENABLED),
-      isEndTestScreenRemovalEnabled: isFeatureEnabled(process.env.FT_END_TEST_SCREEN_REMOVAL_ENABLED),
+      isEndTestScreenRemovalEnabled: Boolean(getArrayOfStrings(process.env.END_TEST_SCREEN_REMOVAL_WHITELIST).length),
```
**suggestion**
```suggestion
      isEndTestScreenRemovalEnabled: getArrayOfStrings(process.env.END_TEST_SCREEN_REMOVAL_WHITELIST).length > 0
```

Encore mieux : le mettre dans une méthode privée `_thereIsAtLeastOneCertificationCenterInWhitelist`

aceol committed:
:recycle: refacto(api) Remove useless inner join

octo-topi commented on api/lib/domain/usecases/retrieve-last-or-create-certification-course.js:
```diff
@@ -41,7 +42,10 @@ module.exports = async function retrieveLastOrCreateCertificationCourse({
   });
 
-  if (featureToggles.isEndTestScreenRemovalEnabled) {
+  const isEndTestRemovalEnabled = await endTestScreenRemovalService.isEndTestScreenRemovalEnabledBySessionId(
```
**nitpick**
Y'a pas mal de redondance dans le nommage, la solution ci-dessous est plus courte
```
if (await endTestScreenRemovalService.bySessionId(sessionid) && !certificationCandidate.isAuthorizedToStart()){
   throw new CandidateNotAuthorizedToJoinSessionError();
}
```
Mais ça fait trop de refacto, et les await dans les `if` c'est un peu :exploding_head: 
Donc pas besoin de le faire, c'est juste une suggestion pour un prochain FT

octo-topi commented on api/tests/integration/infrastructure/repositories/sco-certification-candidate-repository_test.js:
```diff
@@ -63,8 +63,9 @@ describe('Integration | Repository | SCOCertificationCandidate', function () {
       const actualCandidates = candidatesToBeCompared(candidates);
       const expectedCandidates = candidatesToBeCompared(scoCandidates);
-      expect(actualCandidates).to.deep.equal(expectedCandidates);
+      expect(actualCandidates).to.exactlyContain(expectedCandidates);
```
:pray:

octo-topi commented on api/lib/application/preHandlers/end-test-screen-removal-enabled.js:
```diff
@@ -1,8 +1,9 @@
 /* eslint-disable no-restricted-syntax */
```
Si je supprime `eslint-disable`, le lint passe
`npx eslint ./lib/application/preHandlers/end-test-screen-removal-enabled.js`

je propose de supprimer la ligne

aceol replied:
Ah oui, bien joué, on a retiré le probleme (un parseInt)

octo-topi commented on api/lib/domain/errors.js:
```diff
@@ -351,6 +351,14 @@ class CertificationEndedBySupervisorError extends DomainError {
   }
 }
 
+class SupervisorAccessNotAuthorizedError extends DomainError {
```
**nitpick**
`SupervisorAccessNotAuthorizedForCertificationCenterError`
